keys/storage/filestorage/main.go

```diff
@@ -89,11 +89,10 @@ func (s FileStore) Get(name string) (salt []byte, key []byte, info keys.Info, er
 // Info for all keys located in this directory.
 func (s FileStore) List() (keys.Infos, error) {
 	dir, err := os.Open(s.keyDir)
-	defer dir.Close()
-
 	if err != nil {
 		return nil, errors.Wrap(err, "List Keys")
 	}
+	defer dir.Close()
```

Comment:
why do all these defers move? doesnt it mean we can return without closing the dir/f ?

Reply:
the defer will run: see https://blog.golang.org/defer-panic-and-recover
and this part: `Defer statements allow us to think about closing each file right after opening it, guaranteeing that, regardless of the number of return statements in the function, the files will be closed.` immediately after the code block showing the defer after the err check.

Reply:
and see https://github.com/golang/go/issues/3971#issuecomment-66069450

Reply:
@zramsay nice

Reply:
Ah - the dir/f will be nil in the case of err, so there's nothing to close. Thanks.